File: Coding_Java/MergeSort.java
import java.util.Scanner;

class Problem{
    void mergeSort(int[] arr , int beg , int end){
        if(beg>=end)return;
        int mid = beg + (end-beg)/2 ; 
        System.out.println("MergeSort");
        mergeSort(arr , beg , mid );
        System.out.println("MergeSort");
        mergeSort(arr , mid+1 , end );
        System.out.println("Merge");
        merge(arr , beg , mid , end );
    }

    void merge(int[] arr , int beg , int mid , int end){
        int[] left , right ; 
        int l = mid-beg + 1 , r = end -mid ; 
        left = new int[l] ;
        right = new int[r];
        
        for(int i= 0 ; i < l ; i ++){
            left[i] = arr[beg+i];
            System.out.print(left[i]+" ");
        }
        
        for(int i=0 ; i<r ; i++){
            right[i] = arr[mid+1+i];
            System.out.print(right[i]+" ");
        }   

        int lptr=0 , rptr=0 , ptr=beg ; 

        while(lptr < l && rptr<r ){
            if(left[lptr]>right[rptr]){
                arr[ptr] = right[rptr];
                rptr++;
            }
            else{
                arr[ptr] = left[lptr];
                lptr++;
            }
            ptr++;
        }

        while(lptr<l){
            arr[ptr]=left[lptr];
            lptr++;
            ptr++;
        }

        while(rptr<r){
            arr[ptr]=right[rptr];
            rptr++;
            ptr++;
        }

    }

    void displayArr(int[] arr){
        System.out.println("\nPrinting an array:");
        for(int i: arr){
            System.out.print(i);
        }
    }
}

class MergeSort{
    
    
    static final Scanner sc = new Scanner(System.in);
    public static void main(String args[]){
        int[] arr ; 
        int n ; 
        System.out.print("Length:");
        n = sc.nextInt();
        arr = new int[n];
        for(int i=0 ; i< n ; i++){
            System.out.println("Index "+i+":");
            arr[i]=sc.nextInt();
        }
        System.out.println("Arr:"+arr[0]);
        final Problem p = new Problem();
        p.mergeSort(arr, 0,n-1);
        p.displayArr(arr);
        
    }
}

File: Lab/MergeSort.cpp
#include<iostream>
#include<vector>

using namespace std;

void merge(vector<int>& arr , int beg , int mid , int end){
    int l = mid - beg + 1;
    int r = end - mid;
    int left[l+1] , right[r+1];
    for(int i=0 ; i< l ; i++){
        left[i] = arr[beg+i];
    }
    for(int i=0 ; i< r ; i++){
        right[i] = arr[mid+1+i];
    }
    left[l] = __INT_MAX__;
    right[r] = __INT_MAX__;
    int j = 0 , k=0 , i=beg ; 
    while(i<=end){
        if(left[j]<right[k]){
            arr[i] = left[j];
            j++;
        }
        else{
            arr[i] = right[k];
            k++;
        }
        i++;
    }
}

void mergeSort(vector<int>& arr , int beg , int end){
    if (beg<end){
        int mid = beg + (end - beg)/2;
        mergeSort(arr , beg , mid);
        mergeSort(arr , mid+1 , end);
        merge(arr , beg , mid , end);
    }
}

int main(){
    vector<int> arr{7 , 3 , 8 , 5 , 2 , 6};
    mergeSort(arr , 0 , arr.size()-1);
    for(int value:arr){
        cout<<value<<" ";
    }
}